lib/compress/zstdmt_compress.c

```diff
@@ -0,0 +1,686 @@
+        && (zcs->nextJobID <= zcs->doneJobID + zcs->jobIDMask) ) {   /* avoid overwriting job round buffer */
+        size_t const dstBufferCapacity = ZSTD_compressBound(zcs->targetSectionSize);
+        buffer_t const dstBuffer = ZSTDMT_getBuffer(zcs->buffPool, dstBufferCapacity);
+        ZSTD_CCtx* const cctx = ZSTDMT_getCCtx(zcs->cctxPool);
+        unsigned const jobID = zcs->nextJobID & zcs->jobIDMask;
+
+        if ((cctx==NULL) || (dstBuffer.start==NULL)) {
+            zcs->jobs[jobID].jobCompleted = 1;
+            zcs->nextJobID++;
+            ZSTDMT_waitForAllJobsCompleted(zcs);
+            ZSTDMT_releaseAllJobResources(zcs);
+            return ERROR(memory_allocation);
+        }
+
+        DEBUGLOG(1, "preparing job %u to compress %u bytes \n", (U32)zcs->nextJobID, (U32)zcs->targetSectionSize);
+        zcs->jobs[jobID].src = zcs->inBuff.buffer;
+        zcs->jobs[jobID].srcStart = zcs->inBuff.buffer.start;
+        zcs->jobs[jobID].srcSize = zcs->targetSectionSize;
+        zcs->jobs[jobID].params = zcs->params;
+        zcs->jobs[jobID].dict = zcs->nextJobID == 0 ? zcs->dict : NULL;
+        zcs->jobs[jobID].dictSize = zcs->nextJobID == 0 ? zcs->dictSize : 0;
+        zcs->jobs[jobID].fullFrameSize = zcs->frameContentSize;
+        zcs->jobs[jobID].dstBuff = dstBuffer;
+        zcs->jobs[jobID].cctx = cctx;
+        zcs->jobs[jobID].firstChunk = (zcs->nextJobID==0);
+        zcs->jobs[jobID].lastChunk = 0;
+        zcs->jobs[jobID].jobCompleted = 0;
+        zcs->jobs[jobID].dstFlushed = 0;
+        zcs->jobs[jobID].jobCompleted_mutex = &zcs->jobCompleted_mutex;
+        zcs->jobs[jobID].jobCompleted_cond = &zcs->jobCompleted_cond;
+
+        /* get a new buffer for next input - save remaining into it */
+        zcs->inBuff.buffer = ZSTDMT_getBuffer(zcs->buffPool, zcs->inBuffSize);
+        if (zcs->inBuff.buffer.start == NULL) {   /* not enough memory to allocate next input buffer */
+            zcs->jobs[jobID].jobCompleted = 1;
+            zcs->nextJobID++;
+            ZSTDMT_waitForAllJobsCompleted(zcs);
+            ZSTDMT_releaseAllJobResources(zcs);
+            return ERROR(memory_allocation);
+        }
+        zcs->inBuff.filled = (U32)(zcs->inBuffSize - zcs->targetSectionSize);
+        memcpy(zcs->inBuff.buffer.start, (const char*)zcs->jobs[jobID].srcStart + zcs->targetSectionSize, zcs->inBuff.filled);
+
+        DEBUGLOG(3, "posting job %u   (%u bytes)  (note : doneJob = %u=>%u)", zcs->nextJobID, (U32)zcs->jobs[jobID].srcSize, zcs->doneJobID, zcs->doneJobID & zcs->jobIDMask);
+        POOL_add(zcs->factory, ZSTDMT_compressChunk, &zcs->jobs[jobID]);   /* This call is blocking if all workers are busy */
+        zcs->nextJobID++;
+    }
+
+    /* check if there is any data available to flush */
+    {   unsigned const jobID = zcs->doneJobID & zcs->jobIDMask;
+        unsigned jobCompleted;
+        pthread_mutex_lock(&zcs->jobCompleted_mutex);
+        jobCompleted = zcs->jobs[jobID].jobCompleted;
```

Comment:
I think replacing `jobCompleted = zcs->jobs[jobID].jobCompleted` with something like this will work. We make sure that if we can't make forward progress by passing more input, we can make forward progress by flushing output.

```c
while (zcs->jobs[jobID].jobCompleted == 0 && zcs->inBuff.filled == zcs->inBuffSize && zcs->nextJobID > zcs->doneJobID + zcs->jobIDMask) {
    pthread_cond_wait(&zcs->jobCompleted_cond, &zcs->jobCompleted_mutex);
}
jobCompleted = zcs->jobs[jobID].jobCompleted;
```